url294e6c45f3_Ecdosis_NMergeNew_tgz-pJ8-TestJ8/src/edu/luc/nmerge/fastme/tree.java

```diff
@@ -0,0 +1,1616 @@
+		lambda[1] = ((double) b*c + a*d)/((a + c)*(b+d));    
+		lambda[2] = ((double) c*d + a*b)/((a + d)*(b+c));
+	  
+		D_LR = A[e.head.leftEdge.head.index][e.head.rightEdge.head.index];
+		D_LU = A[e.head.leftEdge.head.index][e.tail.index];
+		D_LD = A[e.head.leftEdge.head.index][f.head.index];
+		D_RU = A[e.head.rightEdge.head.index][e.tail.index];
+		D_RD = A[e.head.rightEdge.head.index][f.head.index];
+		D_DU = A[e.tail.index][f.head.index];
+
+		w0 = wf2(lambda[0],D_RU,D_LD,D_LU,D_RD,D_DU,D_LR);
+		w1 = wf2(lambda[1],D_RU,D_LD,D_DU,D_LR,D_LU,D_RD);
+		w2 = wf2(lambda[2],D_DU,D_LR,D_LU,D_RD,D_RU,D_LD);
+		if (w0 <= w1)
+		{
+			if (w0 <= w2) //w0 <= w1,w2
+			{
+				weight[wIndex] = 0.0;
+				return direction.NONE;
+			}
+			else //w2 < w0 <= w1 
+			{
+				weight[wIndex] = w2 - w0;
+				return direction.RIGHT;
+			}
+		}
+		else if (w2 <= w1) //w2 <= w1 < w0
+	    {
+			weight[wIndex] = w2 - w0;
+	      	return direction.RIGHT;
+	    }
+		else //w1 < w2, w0
+	    {
+			weight[wIndex] = w1 - w0;
+			return direction.LEFT;	
+	    }
+	}
+	void NNItopSwitch( edge e, direction d, double[][] A )
+	{
+		edge par, fixed;
+		edge skew, swap;
+	  
+		if ( direction.LEFT == d )
+			swap = e.head.leftEdge;
+		else
+			swap = e.head.rightEdge;
+		skew = e.siblingEdge();
+		fixed = swap.siblingEdge();
+	  	par = e.tail.parentEdge;
+	  	/*perform topological switch by changing f from (u,b) to (v,b)
+	    	and g from (v,c) to (u,c), necessitates also changing parent fields*/
+	  	swap.tail = e.tail;
+	  	skew.tail = e.head;
+	  
+	  	if ( direction.LEFT == d )
+	  		e.head.leftEdge = skew;
+	  	else
+	  		e.head.rightEdge = skew;
+	  	if (skew == e.tail.rightEdge)
+	  		e.tail.rightEdge = swap;
+	  	else
+	  		e.tail.leftEdge = swap;
+	  	//both topsize and bottomsize change for e, but nowhere else
+	  	e.topsize = par.topsize + swap.bottomsize;
+	  	e.bottomsize = fixed.bottomsize + skew.bottomsize;
+	  	NNIupdateAverages( A, e, par, skew, swap, fixed );
+	}
+	void NNIupdateAverages( double[][] A, edge e, edge par, edge skew, 
+		edge swap, edge fixed )
+	{
+		node v;
+		edge elooper;
+		v = e.head;
+		//first, v
+		A[e.head.index][e.head.index] =  
+			(swap.bottomsize* 
+			((skew.bottomsize*A[skew.head.index][swap.head.index]
+			+ fixed.bottomsize*A[fixed.head.index][swap.head.index]) 
+			/ e.bottomsize) +
+			par.topsize*
+			((skew.bottomsize*A[skew.head.index][par.head.index]
+			+ fixed.bottomsize*A[fixed.head.index][par.head.index]) 
+			/ e.bottomsize)
+			) / e.topsize; 
+		// next, we loop over all the edges 
+		// which are below e
+		elooper = e.findBottomLeft(); 
+		while ( e != elooper )  
+		{
+			A[e.head.index][elooper.head.index] = 
+				A[elooper.head.index][v.index] 
+		        = (swap.bottomsize*A[elooper.head.index][swap.head.index] +
+	   	        par.topsize*A[elooper.head.index][par.head.index]) 
+	            / e.topsize;
+                elooper = depthFirstTraverse(elooper);
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 1119:1127, 1131:1139.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.